#376: View location expander for Controller Views and Components.

jtkech opened the pull request:
Fixes #376 


- We use `extension.SubPath` which contains `Modules or Themes` or another extension location..


- As an example, for a view component defined as below.

        \Modules\Orchard.ContentTypes\Views\Shared\Components\SelectContentTypes\Default.cshtml

- Can be overridden by one of the followings (through a module with higher priority or which depend on the above module, or by a theme if implicit dependencies are implemented)

      \Modules\Orchard.Demo\Views\Orchard.ContentTypes\Admin\Components\SelectContentTypes\Default.cshtml
      \Modules\Orchard.Demo\Views\Orchard.ContentTypes\Shared\Components\SelectContentType\Default.cshtml

- Here we assume we can override a controller view or component (not shapes) only through a main feature (which have the same id as its extension). Let me know if it is right.

- Otherwise, we would need to select all features, and, because sub features have different ids, dependencies and priorities, we would need others paths with segments related to sub feature ids

Best.

jtkech commented:
- Not so simple, we need to only use the list of enabled extensions (still ordered) and the current theme and its base theme(s). As done in the shape table manager with `IsEnabledModuleOrRequestedTheme`.

- So, we need to resolve services on the fly with the `ViewLocationExpanderContext` context from which we can get `context.ActionContext.HttpContext.RequestServices`. Working on it.

- Otherwise it works, i can override a controller view / view component from a module / theme.

- UPDATE: Done for enabled extensions (still ordered). Todo: get the selected theme (and base theme).

- UPDATE: as said, here we assume we can override a view / component (not shapes) only through a main feature (same id as its extension). Seems to be the case in O1 where all enabled features are grabbed but we select their extensions on which a Distinct() is done, so seems to be the same.

By doing this, i found some mismatches, e.g in O1 in `ShapeTemplateBindingStrategy` we use.

            // all features which are ordered by dependencies and priorities
            var availableFeatures = _extensionManager.AvailableFeatures();
            // then keep only those which are enabled
            var activeFeatures = availableFeatures.Where(FeatureIsEnabled);
            ...

And in O2.

           // get directly enabled features but which are not ordered 
           var enabledFeatures = _shellFeaturesManager.GetEnabledFeaturesAsync().Result
            ...

So, not exactly the same. But here i think that the real mismatch is that GetEnabledFeaturesAsync() doesn't return an ordered list. For each feature, dependencies are ordered but not the final collection. In O1 the EnabledFeatures() extension return an ordered list.

Here not so important because things are re-ordered in shape table manager, and i don't know if it's worth to check this everywhere.

jtkech commented:
So, it works well and here the logic but in a preview way because some things need to be discussed. After, maybe better to implement a service and cache part of things e.g per tenant / theme.

- To get the theme i needed to use `IThemeManager` but this add to `...Mvc.Modules.Hosting` a dependency on `Orchard.DisplayManagement` which is not good, but here we work on views so ...

- We already seen the theme implicit dependencies on all other modules and its base theme(s). We need to see how to implement this, in each specific case as done right now in shape table manager or at a lower level. I say this because here there is another use case (see below).

- As seen in O1 in `ThemeAwareViewEngine.cs`, there are other specific theme dependencies. All themes are considered as a base theme of the admin theme! Here the comments.

      // Special case: conceptually, the base themes of "TheAdmin" is the list of all enabled themes.
      // This is so that any enabled theme can have controller/action/views in the Admin of the site.

- I've implemented the above as in O1, but in this code i don't like the use of the hard coded string `TheAdmin`. But to use the `AdminThemeService` we would need a dependency on `Orchard.Admin`!

- Note: we grab extensions from main features (same id as their extension), in O1 it's quite the same but by using a `Distinct()` on extensions and with this comment: `// is Distinct guaranty to keep order?`. I think it's now better if we assume that these are main features which provide physical views paths.

- In O1 enabledFeatures are ordered by dependencies, but not in O2 through `GetEnabledFeatures()`. All dependencies of an individual feature are ordered, but not the final collection. **But maybe there is a good reason**. Note: in `CompositionStrategy` the same code  is used but maybe for another job.

      GetEnabledFeatures(...) {
      //return shellDescriptor.Features.SelectMany(shellFeature => extensionManager.GetFeatureDependencies(shellFeature.Id)).Distinct();

- So here, i've removed the above line and used the following in `GetEnabledFeatures()`, as in O1. This because `GetExtensions().Features` is ordered (as `AvailableFeatures()` in O1).

      GetEnabledFeatures(...) {
      // enabled features ordered by dependencies and priorities as in O1
      return extensionManager.GetExtensions().Features.Where(f => shellDescriptor.Features.Any(sf => sf.Id == f.Id));

- As seen in O1 in `ThemeAwareViewEngine.cs`, do we need to do something for `PartialViews`?

Best.

Jetski5822 commented on src/Microsoft.AspNetCore.Mvc.Modules.Hosting/Mvc/Razor/ModuleViewLocationExpander.cs:
```diff
@@ -16,8 +23,23 @@ public void PopulateValues(ViewLocationExpanderContext context)
+            var enabledFeatures = shellFeaturesManager.GetEnabledFeaturesAsync().Result;
+
+            var features = availableFeatures.Where(af => enabledFeatures.Any(ef => af.Id == ef.Id));
+            var extensions = features.Where(f => f.Id == f.Extension.Id).Select(f => f.Extension).Reverse();
```
not sure this is accurate here... Where(f => f.Id == f.Extension.Id), do we need it?

jtkech replied:
We grab extensions from main features (not from sub features) to keep the order based on these main features which provide through their extension physical view subpaths.

Note: just commited my last updates.

jtkech commented:
This is a preview, things need to be discussed (see my previous comments)
Need to leave, come back in 3 or 4 hours.

sebastienros commented:
The dependency on ITheme is only because the order is the dependency is not respected, right ?

jtkech commented:
No, because as for shape table we need to know the selected theme (not only enabled themes). moreover there is here another implicit dependencies applied to the admin theme (as i've seen in O1) (see above).

Really need to leave but come back later.

sebastienros commented:
There is a mismatch here, the ViewLocationExpander is called once per compiled view, while the theme selection is called per request.
Also each tenant has a different set of modules (default enabled feature of the modules) that should also enable the views for this module, so I assume we should have multiple sets of view locations, or is there a way to have only one if we change the view engine itself instead of these options. Or do we need as many MVC middleware and configuration per tenant.

sebastienros commented:
We currently have a custom FileProvider that resolves the _Layout.cshtml template automatically. What about extending the idea to have the FileProvider (which is DI injected) to look at the current Theme to return the views. We could just use a special case file url like `/CurrentTheme/Views/{0} ` and when the FileProvider sees this pattern it would transform the file path.

jtkech commented:
- Yes, i've seen the mismatch where ViewLocationExpander is called once per compiled view.

- I also thought about a service which cache views locations per tenant and per theme (as the shape table manager). Then when views locations are requested it returns the right list based on the current theme.

- A file provider seems a good idea, i'll think about this. Oh, just seen the `ThemingFileProvider` class.

- Oh, this would solve the dependency we needed to have on `Orchard.DisplayManagement`. Right?

- But what about the special case of the admin theme. I assume that the admin theme id will not be always `TheAdmin`. **So, how to know the current admin theme id without having a dependency on `Orchard.Admin`?**

- Oh, maybe another file provider based on the current admin theme and implemented by `Orchard.Admin`. **Or a global controller views and components location provider implemented by another module** which would need to depend on `Orchard.Admin` and `Orchard.DisplayManagement` to be able to get the current theme and the admin theme name.

I lost myself :)

sebastienros commented:
You don't need to know about TheAdmin, just resolve the current theme, if you are in an admin view, then it will be `TheAdmin`.

jtkech commented:
Yes, but when with the admin theme i've see in O1 other specific implicit dependencies where all themes are considered as a base theme of the admin (as said in my previous comments).

Here the comments grabbed from O1

    // Special case: conceptually, the base themes of "TheAdmin" is the list of all enabled themes.
    // This is so that any enabled theme can have controller/action/views in the Admin of the site.

jtkech commented:
Before leaving what i've quickly done.

> There is a mismatch here, the ViewLocationExpander is called once per compiled view, while the theme selection is called per request. Also each tenant has a different set of modules ...

Indeed it's a big problem. What i've seen is that `PopulateValues()` is always called and where we can update contextual values which are used in provided paths. But this doesn't seem to be sufficient.

Need to leave but will check it this night.

Meanwhile what i've quickly done

- I've moved `ModuleViewLocationExpander` in `Orchard.DisplayManagement` and renamed it `ThemeAwareViewLocationExpander`.

- So, we can use `IThemeManager` without adding extra dependencies in project.json

- And we use `ISiteService` to get the admin theme name to check if it is the current theme and implement the specific dependency where each theme is considered as a base theme of the admin theme (as implemented in O1 in `ThemeAwareViewEngine`).

- Seems to work and it adds an overhead of 15 ms the 1st time it is executed. Then after it is fast because things are cached in the related services.

jtkech commented:
@sebastienros coming back to this PR i think i have a solution for the following.

> There is a mismatch here, the ViewLocationExpander is called once per compiled view, while the theme selection is called per request. Also each tenant has a different set of modules ...

Indeed it's a big problem. But, as said in my previous comment, `PopulateValues()`, where we can update contextual values, is always called ...

Just tried it, if in `PopulateValues()` we change a context value, then `ExpandViewLocations()` is re-called. So, if we use context values related to the theme and tenant, i think it should work.

Will do and try it when i have time.

sebastienros commented:
Yes, talked to @pranakm and from his explanation it's exactly `PopulateValues` that we need. Just pass the name of the current theme and the view locations will then vary by that.

sebastienros commented:
And the tenant name too ... as it will impact the list of available modules.

jtkech commented:
Yes, just tried it, it works. When i change the theme / tenant views locations are updated.

Tomorrow i'll rebase the PR on last master.

sebastienros commented:
Try to enable a module with views, and see that the views are found.

jtkech commented:
Yes, a little problem here. When you enable / disable a module because we still use the same context values views locations are not updated.

And when we return back to the front end, we swith the theme value from TheAdmin to TheTheme, but because this value was previously used, views location are not updated.

jtkech commented:
When we switch the theme / tenant, one idea would be to provide e.g a random value just to trigger `ExpandViewLocations()`. But i think the razor engine will cache unecessary views locations.

Because we re-start a tenant after enabling / disabling a module, we would need to register and configure a RazorViewEngine at the tenant level. Will see tomorrow.

sebastienros commented:
You don't need to care about enabling/disabling modules, I striked that earlier. We already add all features to the list to find the views. Only the theme will actually override the MVC views, shapes are handled differently and already take care of the enabled state. We just need to list all views from all modules to be able to find the View Component views, and have specific locations for each theme as they are the ones to override the controller views.

jtkech commented:
Not sure i fully understand

So here it's to allow a theme / module to override a controller view / view component of another module.

> Only the theme will actually override the MVC views

So here you mean that a module will not be able to override a controller view of another module. Only a theme can do this. Right?

And it's the same for views components. Right?

So we just need that all modules through their own controller views (or overrriden by the theme) will be able to find their components views or those overriden by the curent theme. Not sure i'm clear here!

jtkech commented:
And if i well understand we don't need anymore a context value for the tenant, only for the theme.

sebastienros commented:
Right.

On Dec 7, 2016 7:30 PM, "jtkech" <notifications@github.com> wrote:

> And if i well understand we don't need anymore a context value for the
> tenant, only for the theme.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/Orchard2/pull/386#issuecomment-265643905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHJ7c_7ZxlbaAEyUMegI2f1M63jkXX0ks5rF3nAgaJpZM4LDthY>
> .
>

jtkech commented:
Done, will do more tests.

- Resolve MVC view / component paths for each extension and according to the current theme.

- The current theme and its base themes can override an MVC view / component of an extension.

- In admin, the admin theme and all enabled themes can override an MVC view / component of an extension. Here, all themes are considered as a baseTheme of the admin theme.

    as an example to override a view component (if not already overriden in the admin theme).

      \Themes\TheTheme\Views\Orchard.ContentTypes\Admin\Components\SelectContentTypes\Default.cshtml
      \Themes\TheTheme\Views\Orchard.ContentTypes\Shared\Components\SelectContentType\Default.cshtml
- In fact now, here we only need the implicit dependency of a theme on its base themes. And we normally implement locally a specific dependency related to themes when we are in the admin (see above).

- So, because @Jetski5822 first work on explicit dependencies, i think i don't need to rebase on master.

- Note: Unit tests pass on my machine.

Best.

Jetski5822 commented:
Hey did my changes help you?

Sent from my iPhone

> On 8 Dec 2016, at 15:12, jtkech <notifications@github.com> wrote:
> 
> Done, will do more tests.
> 
> Resolve MVC view / component paths for each extension and according to the current theme.
> 
> The current theme and its base themes can override an MVC view / component of an extension.
> 
> In admin, the admin theme and all enabled themes can override an MVC view / component of an extension. Here, all themes are considered as a baseTheme of the admin theme.
> 
> as an example to override a view component (if not already overriden in the admin theme).
> 
> \Themes\TheTheme\Views\Orchard.ContentTypes\Admin\Components\SelectContentTypes\Default.cshtml
> \Themes\TheTheme\Views\Orchard.ContentTypes\Shared\Components\SelectContentType\Default.cshtml
> In fact now, here we only need the implicit dependency of a theme on its base themes. And we normally implement locally a specific dependency related to themes when we are in the admin (see above).
> 
> So, because @Jetski5822 first work on explicit dependencies, i think i don't need to rebase on master.
> 
> Note: Unit tests pass on my machine.
> 
> Best.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

pranavkm commented on src/Orchard.DisplayManagement/Razor/ThemeAwareViewLocationExpander.cs:
```diff
@@ -17,36 +17,47 @@ public class ThemeAwareViewLocationExpander : IViewLocationExpander
         public void PopulateValues(ViewLocationExpanderContext context)
         {
+            var themeManager = context.ActionContext.HttpContext.RequestServices.GetService<IThemeManager>();
+            context.Values["Theme"] = themeManager.GetThemeAsync().Result.Id;
```
`themeManager.GetThemeAsync().GetAwaiter().Result.Id` in case ThemeManager throws. See http://stackoverflow.com/a/17284612/235591

jtkech commented:
@Jetski5822 yes thanks, it already helped me because now we know what will be implemented internally and what we have to implement locally. But not yet tested because this PR is not rebased on master. I will do it this night when i have time.

I've seen that you began with explicit dependencies and you plan to implement the BaseTheme implicit dependency. So, you will only have to add the implicit dependency of a theme on all modules, as described in the piece of code provided in #389.

@pranavkm thanks for your review, i've replaced `.Result` with `.GetAwaiter().GetResult()` in 3 places to prevent from having `AggregateException` with only generic message.

@sebastienros does it means that it would be better to do the same everywhere we use `.Result`?

Best.

sebastienros commented:
I assume it's safe to use .GetResult() as we are on Core. And we should also look at where we have AggregateExceptions handled, just in case we had these because of a .Result that could throw. (Shape rendering for instance)

jtkech commented:
Yes, i will check this.

Just rebased this PR on master (only locally) because anyway i need to replace `extensionManager.GetExtensions().Features` with `extensionManager.GetFeatures()`.

But right now i can't get the master branch working. I'll try to see why, seems that in `CompositionStrategy.ComposeAsync()` we try to load features by using somewhere a `FeaturesProvider` which is not yet registered (through Orchard.Commons startup) ...

jtkech commented:
@sebastienros 

Just got the master branch working, i needed to also register the new `FeaturesProvider` at the host level and use `TryAddScoped` when through `Orchard.Commons` startup, as before with the `FeatureManager`.

Then i needed to solve a StackOverflow when `GetFeatures()` call `GetDependentFeatures()` which call `GetFeatures()` ... I will give more details on what i've done which need to be reviewed.

I need to check what i've done and maybe have a thought on implicit dependencies.

**When ready to be reviewed do i need to open a new PR or can i push the changes here?**

sebastienros commented:
here

jtkech commented:
Ok

Jetski5822 commented:
Oh boy, good spot, sorry about that - thanks for fixing

Sent from my iPhone

> On 9 Dec 2016, at 08:54, jtkech <notifications@github.com> wrote:
> 
> Ok
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jtkech commented:
@Jetski5822 no worries, then i can see where `OrderByDependenciesAndPriorities()` is used in the new structure. I've seen others little things i'm working on.

I will propose a fix here with more details.

Best.

Jetski5822 commented:
@jtkech I fixed it on master for you man.... I am going to look at optimizing the cache in Extension Manager next. The API wont change.

jtkech commented:
@Jetski5822 cool, thanks. So i will rebase this one.

So, there are 2 or 3 things i would like to discuss with you. Maybe easier to have a talk on skype.

jtkech commented:
@Jetski5822 

So, i've done here a temporary commit with some suggestions. Please review, all are in the comment lines, particularly about implicit dependencies where i've done a temporary code just to show what we need.

I had no time to implement it in the right way but i have some ideas to do it without having to make extensions aware of theming (see my comments in the code).

Then, if you agree i will remove comments update the code according to your feedback. Then if merged on master you will be able to finalize it.

Best.

jtkech commented:
@Jetski5822 

- So, cleanup done
- Lazy for all concurrent dictionaries.
- I had already checked that ordering according to explicit dependencies works.
- Idem for the new implicit dependency on BaseTheme implemented internally.

So, it remains the implicit dependency of a theme on all modules. As discussed with some HasDependency (or another name) based on the FeatureInfo type. Maybe an array of func where each builder can add its custom one. Then, when executed, we have a dependency as soon one of them return true.

Good also to expose it through the FeatureInfo interface, then we can use it e.g when we order things in the shape table manager, this without rewriting code.

Best.

Jetski5822 commented:
@jtkech sweet - so we just need to move HasDependency outside of the Extensions code. Will update master and let you know when its done.
